pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/exchange/BlockExchange.java

```diff
@@ -58,23 +68,44 @@ public static BlockExchange getExchange(List<SendingMailbox> sendingMailboxes, R
     }
   }
 
-  protected BlockExchange(List<SendingMailbox> sendingMailboxes, BlockSplitter splitter) {
+  protected BlockExchange(OpChainExecutionContext context, List<SendingMailbox> sendingMailboxes,
+      BlockSplitter splitter) {
+    _context = context;
     _sendingMailboxes = sendingMailboxes;
     _splitter = splitter;
+    _context.getMailboxService().submitExchangeRequest(this);
   }
 
-  public void send(TransferableBlock block)
+  public boolean send(TransferableBlock block)
       throws Exception {
-    if (block.isEndOfStreamBlock()) {
-      for (SendingMailbox sendingMailbox : _sendingMailboxes) {
-        sendBlock(sendingMailbox, block);
+    _queue.offer(block, _context.getTimeoutMs(), TimeUnit.MILLISECONDS);
+    return _queue.remainingCapacity() > 0;
+  }
+
+  public TransferableBlock sendBlock() {
+    try {
+      TransferableBlock block = _queue.poll(_context.getTimeoutMs(), TimeUnit.MILLISECONDS);
+      // TODO: make it into a formal callback in constructor
+      _context.getMailboxService().getCallback().accept(_context.getId());
+      if (block == null) {
+        return TransferableBlockUtils.getErrorTransferableBlock(
+            new TimeoutException("Timed out while sending data on: " + _context.getId()));
       }
-      return;
+      if (block.isEndOfStreamBlock()) {
+        for (SendingMailbox sendingMailbox : _sendingMailboxes) {
+          sendBlockToMailbox(sendingMailbox, block);
+        }
+        return block;
+      }
+      route(_sendingMailboxes, block);
+      return block;
+    } catch (Exception e) {
+      return TransferableBlockUtils.getErrorTransferableBlock(
+          new TimeoutException("Exception while sending data via mailbox on: " + _context.getId()));
     }
```

Comment:
currently this is the one that ensures the sending thread guarantees to return within the timeout if anything occurs.